mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxFactory.kt

```diff
@@ -1,10 +1,113 @@
 package com.airbnb.mvrx
 
+import android.os.Bundle
+import android.os.Parcelable
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.ViewModelProvider
 import androidx.annotation.RestrictTo
+import java.io.Serializable
+import java.lang.IllegalStateException
+import kotlin.reflect.full.primaryConstructor
 
 @RestrictTo(RestrictTo.Scope.LIBRARY)
-class MvRxFactory<V>(private val factory: (Class<*>) -> V) : ViewModelProvider.Factory {
-    @Suppress("UNCHECKED_CAST") override fun <T : ViewModel?> create(modelClass: Class<T>) = factory(modelClass) as T
-}
\ No newline at end of file
+
+class MvRxFactory<VM : BaseMvRxViewModel<S>, S : MvRxState>(
+    private val viewModelClass: Class<out VM>,
+    private val stateClass: Class<out S>,
+    private val viewModelContext: ViewModelContext,
+    private val key: String,
+    private val forExistingViewModel: Boolean,
+    private val initialStateFactory: MvRxStateFactory<VM, S>
+) : ViewModelProvider.Factory {
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+        val (viewModelContext, viewModel) = viewModelContext.savedStateRegistry.consumeRestoredStateForKey(key)?.let { savedStateBundle ->
+            val restoredArgs = savedStateBundle.get(KEY_MVRX_SAVED_ARGS)
+            val restoredState = savedStateBundle.getBundle(KEY_MVRX_SAVED_INSTANCE_STATE)
+
+            requireNotNull(restoredState) { "State was not saved prior to restoring!" }
+
+            val restoredContext = when (viewModelContext) {
+                is ActivityViewModelContext -> ActivityViewModelContext(viewModelContext.activity, restoredArgs)
+                is FragmentViewModelContext -> FragmentViewModelContext(viewModelContext.activity, restoredArgs, viewModelContext.fragment)
+            }
+            restoredContext to createViewModel(viewModelClass, stateClass, restoredContext, key, restoredState::restorePersistedState)
+        } ?: viewModelContext to createViewModel(viewModelClass, stateClass, viewModelContext, key, forExistingViewModel = forExistingViewModel, initialStateFactory = initialStateFactory)
+
+        viewModelContext.savedStateRegistry.registerSavedStateProvider(key) {
+            withState(viewModel) { state ->
+                Bundle().apply {
+                    putBundle(KEY_MVRX_SAVED_INSTANCE_STATE, state.persistState())
+                    when (val args = viewModelContext.args) {
+                        is Parcelable -> putParcelable(KEY_MVRX_SAVED_ARGS, args)
+                        is Serializable -> putSerializable(KEY_MVRX_SAVED_ARGS, args)
```

Comment:
can you throw in the else case? better to fail and detect it, then fail without knowing I think